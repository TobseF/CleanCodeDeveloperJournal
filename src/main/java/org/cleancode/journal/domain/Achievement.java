package org.cleancode.journal.domain;

import java.util.Objects;

public class Achievement {


    private Group group;
    /**
     * Unique id
     */
    private String id;
    /**
     * XP are points you receive for completing achievements.
     */
    private int experience;
    /**
     * CHA: Gained by completing achievements that involve social skills, interaction or communication with others.
     */
    private int charisma;
    /**
     * TAL: Gained by completing achievements that improve or require a specialized skill or ability.
     */
    private int talent;
    /**
     * STR: Gained by completing achievements that improve or require physical health and fitness.
     */
    private int strength;
    /**
     * INT: Gained by completing achievements that improve knowledge or requires education and research to complete-
     */
    private int intellect;

    public Group getGroup() {
        return group;
    }

    public void setGroup(Group group) {
        this.group = group;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public int getExperience() {
        return experience;
    }

    public void setExperience(int experience) {
        this.experience = experience;
    }

    public int getCharisma() {
        return charisma;
    }

    public void setCharisma(int charisma) {
        this.charisma = charisma;
    }

    public int getTalent() {
        return talent;
    }

    public void setTalent(int talent) {
        this.talent = talent;
    }

    public int getStrength() {
        return strength;
    }

    public void setStrength(int strength) {
        this.strength = strength;
    }

    public int getIntellect() {
        return intellect;
    }

    public void setIntellect(int intellect) {
        this.intellect = intellect;
    }

    @Override
    public String toString() {
        return "Achievement{" + id + '\'' + ", XP " + experience + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Achievement that = (Achievement) o;
        return experience == that.experience &&
                charisma == that.charisma &&
                talent == that.talent &&
                strength == that.strength &&
                intellect == that.intellect &&
                group == that.group &&
                id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(group, id, experience, charisma, talent, strength, intellect);
    }

    public enum Group {
        Grade, Log, CleanCode, Social, Dev, Health, Knowledge;
    }
}
