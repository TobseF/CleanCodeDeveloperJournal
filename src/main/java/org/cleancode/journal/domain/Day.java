package org.cleancode.journal.domain;

import org.cleancode.journal.util.ToStringUtil;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Day implements Comparable<Day> {

    private final LocalDate date;
    private List<LogEntry> logEntries = new ArrayList<>();

    public Day(LocalDate date) {
        this.date = date;
    }

    public LocalDate getDate() {
        return date;
    }

    public void addLogEntry(LogEntry logEntry) {
        logEntries.add(logEntry);
    }

    public Type getType() {
        if (logEntries.isEmpty()) {
            return Type.Empty;
        } else if (logEntries.stream().allMatch(LogEntry::isFulfilled)) {
            return Type.Fulfilled;
        }
        return Type.PartialFulfilled;
    }

    @Override
    public int compareTo(Day o) {
        return date.compareTo(o.date);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Day day = (Day) o;
        return date.equals(day.date) && logEntries.equals(day.logEntries);
    }

    @Override
    public int hashCode() {
        return Objects.hash(date, logEntries);
    }

    @Override
    public String toString() {
        return ToStringUtil.toString(date) + ":" + logEntries.size();
    }

    public enum Type {Empty, Fulfilled, PartialFulfilled}


}
