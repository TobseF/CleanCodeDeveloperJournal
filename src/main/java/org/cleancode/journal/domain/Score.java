package org.cleancode.journal.domain;

import java.io.Serializable;
import java.util.stream.IntStream;

public class Score implements Serializable {

    /**
     * XP are points you receive for completing achievements.
     */
    private int experience;
    /**
     * CHA: Gained by completing achievements that involve social skills, interaction or communication with others.
     */
    private int charisma;
    /**
     * TAL: Gained by completing achievements that improve or require a specialized skill or ability.
     */
    private int talent;
    /**
     * STR: Gained by completing achievements that improve or require physical health and fitness.
     */
    private int strength;
    /**
     * INT: Gained by completing achievements that improve knowledge or requires education and research to complete-
     */
    private int intellect;

    public int getExperience() {
        return experience;
    }

    public int getMax() {
        return IntStream.of(charisma, talent, strength, intellect).max().orElse(0);
    }

    public void setExperience(int experience) {
        this.experience = experience;
    }

    public int getCharisma() {
        return charisma;
    }

    public void setCharisma(int charisma) {
        this.charisma = charisma;
    }

    public int getTalent() {
        return talent;
    }

    public void setTalent(int talent) {
        this.talent = talent;
    }

    public int getStrength() {
        return strength;
    }

    public void setStrength(int strength) {
        this.strength = strength;
    }

    public int getIntellect() {
        return intellect;
    }

    public void setIntellect(int intellect) {
        this.intellect = intellect;
    }


    public void add(Score score) {
        charisma += score.charisma;
        intellect += score.intellect;
        talent += score.talent;
        strength += score.strength;
        experience += score.experience;
    }

    public String getSkills() {
        String skills = "";
        if (charisma != 0) {
            skills += "CHA " + charisma + " ";
        }
        if (intellect != 0) {
            skills += "INT " + intellect + " ";
        }
        if (talent != 0) {
            skills += "TAL " + talent + " ";
        }
        if (strength != 0) {
            skills += "STR" + strength + " ";
        }
        return skills;
    }

    public Score clone() {
        Score score = new Score();
        score.charisma = charisma;
        score.experience = experience;
        score.intellect = intellect;
        score.talent = talent;
        score.strength = strength;
        return score;
    }
}
