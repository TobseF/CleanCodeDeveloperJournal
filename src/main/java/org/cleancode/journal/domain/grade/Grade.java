package org.cleancode.journal.domain.grade;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;

public class Grade implements Serializable {

    private Locale locale;
    private GradeColor gradeColor;
    private int number;

    public List<GradeTopic> listAllGradeTopics() {
        return Stream.concat(principles.stream(), practices.stream()).collect(toList());
    }

    private List<Principle> principles = new ArrayList<>();
    private List<Practice> practices = new ArrayList<>();

    public Locale getLocale() {
        return locale;
    }

    public void setLocale(Locale locale) {
        this.locale = locale;
    }

    public GradeColor getGradeColor() {
        return gradeColor;
    }

    public void setGradeColor(GradeColor gradeColor) {
        this.gradeColor = gradeColor;
    }

    public int getNumber() {
        return number;
    }

    public void setNumber(int number) {
        this.number = number;
    }

    public List<Principle> getPrinciples() {
        return principles;
    }

    public void setPrinciples(List<Principle> principles) {
        this.principles = principles;
    }

    public void addPrinciples(Principle principle) {
        this.principles.add(principle);
    }

    public List<Practice> getPractices() {
        return practices;
    }

    public void setPractices(List<Practice> practices) {
        this.practices = practices;
    }

    public void addPractice(Practice practice) {
        this.practices.add(practice);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Grade grade = (Grade) o;
        return number == grade.number && locale.equals(grade.locale) && gradeColor == grade.gradeColor && principles.equals(grade.principles) && practices.equals(grade.practices);
    }

    @Override
    public int hashCode() {
        return Objects.hash(locale, gradeColor, number, principles, practices);
    }

    @Override
    public String toString() {
        return gradeColor.toString();
    }
}
