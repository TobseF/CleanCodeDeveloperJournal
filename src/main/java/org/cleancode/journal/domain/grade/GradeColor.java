package org.cleancode.journal.domain.grade;

import java.io.Serializable;

public enum GradeColor implements Serializable {
    Black, Red, Orange, Yellow, Green, Blue, White;

    public int getNumber() {
        return ordinal();
    }
}
