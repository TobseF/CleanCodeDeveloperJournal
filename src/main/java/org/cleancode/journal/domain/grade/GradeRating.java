package org.cleancode.journal.domain.grade;

import java.util.Objects;

public class GradeRating {

    private StarsOutOf3 evolvability = StarsOutOf3._0;
    private StarsOutOf3 correctness = StarsOutOf3._0;
    private StarsOutOf3 productionEfficiency = StarsOutOf3._0;
    private StarsOutOf2 continuousImprovement = StarsOutOf2._0;
    private Responsibility responsibility;

    public StarsOutOf3 getEvolvability() {
        return evolvability;
    }

    public void setEvolvability(StarsOutOf3 evolvability) {
        this.evolvability = evolvability;
    }

    public StarsOutOf3 getCorrectness() {
        return correctness;
    }

    public void setCorrectness(StarsOutOf3 correctness) {
        this.correctness = correctness;
    }

    public StarsOutOf3 getProductionEfficiency() {
        return productionEfficiency;
    }

    public void setProductionEfficiency(StarsOutOf3 productionEfficiency) {
        this.productionEfficiency = productionEfficiency;
    }

    public StarsOutOf2 getContinuousImprovement() {
        return continuousImprovement;
    }

    public void setContinuousImprovement(StarsOutOf2 continuousImprovement) {
        this.continuousImprovement = continuousImprovement;
    }

    public Responsibility getResponsibility() {
        return responsibility;
    }

    public void setResponsibility(Responsibility responsibility) {
        this.responsibility = responsibility;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        GradeRating that = (GradeRating) o;
        return evolvability == that.evolvability && correctness == that.correctness && productionEfficiency == that.productionEfficiency && continuousImprovement == that.continuousImprovement && responsibility == that.responsibility;
    }

    @Override
    public int hashCode() {
        return Objects.hash(evolvability, correctness, productionEfficiency, continuousImprovement, responsibility);
    }

    public enum Responsibility {Team, SingleDev}

    public enum StarsOutOf2 {_0, _1, _2}

    public enum StarsOutOf3 {_0, _1, _2, _3}
}
