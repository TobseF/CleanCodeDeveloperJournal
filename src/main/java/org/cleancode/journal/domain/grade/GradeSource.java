package org.cleancode.journal.domain.grade;

import java.io.Serializable;
import java.util.Objects;

public class GradeSource implements Serializable {
    private String source;
    private String author;
    private String description;

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getSource() {
        return source;
    }

    public void setSource(String source) {
        this.source = source;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        GradeSource that = (GradeSource) o;
        return source.equals(that.source) && author.equals(that.author) && description.equals(that.description);
    }

    @Override
    public int hashCode() {
        return Objects.hash(source, author, description);
    }


}
