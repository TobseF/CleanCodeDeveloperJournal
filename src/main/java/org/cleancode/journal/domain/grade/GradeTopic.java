package org.cleancode.journal.domain.grade;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class GradeTopic {

    private String name;
    private String id;
    private String sectionWhy;
    private String description;
    private GradeRating gradeRating;

    private List<GradeSource> sources = new ArrayList<>();

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getSectionWhy() {
        return sectionWhy;
    }

    public void setSectionWhy(String sectionWhy) {
        this.sectionWhy = sectionWhy;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public GradeRating getGradeRating() {
        return gradeRating;
    }

    public void setGradeRating(GradeRating gradeRating) {
        this.gradeRating = gradeRating;
    }

    public List<GradeSource> getSources() {
        return sources;
    }

    public void setSources(List<GradeSource> sources) {
        this.sources = sources;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof GradeTopic)) return false;
        GradeTopic that = (GradeTopic) o;
        return name.equals(that.name) && id.equals(that.id) && Objects.equals(sectionWhy, that.sectionWhy) && Objects.equals(description, that.description) && Objects.equals(gradeRating, that.gradeRating) && sources.equals(that.sources);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, id, sectionWhy, description, gradeRating, sources);
    }
}
