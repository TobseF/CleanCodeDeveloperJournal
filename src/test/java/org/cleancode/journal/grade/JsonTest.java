package org.cleancode.journal.grade;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.approvaltests.Approvals;
import org.cleancode.journal.domain.Achievement;
import org.cleancode.journal.domain.Score;
import org.cleancode.journal.domain.grade.*;
import org.cleancode.journal.domain.grade.GradeRating.Responsibility;
import org.cleancode.journal.domain.grade.GradeRating.StarsOutOf3;
import org.junit.Test;

import java.util.Locale;

public class JsonTest {

    private Gson gson = new GsonBuilder().setPrettyPrinting().create();

    @Test
    public void gradeTest() throws Exception {

        Grade grade = new Grade();
        grade.setLocale(Locale.ENGLISH);
        grade.setGradeColor(GradeColor.Red);
        grade.setNumber(GradeColor.Red.getNumber());

        Practice practice = new Practice();
        grade.addPractice(practice);
        practice.setName("Boy Scout Rule");
        practice.setId("boy-scout-rule");
        practice.setDescription("Lorem Ispum");
        practice.setSectionWhy("Because of you");
        {
            GradeRating rating = new GradeRating();
            rating.setCorrectness(StarsOutOf3._2);
            rating.setResponsibility(Responsibility.SingleDev);
            practice.setGradeRating(rating);
        }

        Principle principle = new Principle();
        grade.addPrinciples(principle);
        principle.setName("Keep it simple, stupid (KISS)");
        principle.setId("kiss");
        principle.setDescription("Lorem Ispum");
        principle.setSectionWhy("Because you can");

        {
            GradeRating rating = new GradeRating();
            rating.setEvolvability(StarsOutOf3._2);
            rating.setCorrectness(StarsOutOf3._2);
            rating.setResponsibility(Responsibility.SingleDev);
            principle.setGradeRating(rating);
        }


        Approvals.verifyJson(gson.toJson(grade));
    }

    @Test
    public void achievementTest() {
        Achievement achievement = new Achievement();
        achievement.setGroup(Achievement.Group.Dev);
        achievement.setId("achievement.dev.write-test");
        Score score = achievement.getScore();
        score.setCharisma(42);
        score.setExperience(43);
        score.setStrength(44);
        score.setIntellect(45);
        score.setTalent(46);

        Achievement[] achievements = {achievement, achievement};

        Approvals.verifyJson(gson.toJson(achievements));
    }
}