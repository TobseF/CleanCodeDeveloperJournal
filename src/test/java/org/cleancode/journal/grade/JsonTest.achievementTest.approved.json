[
  {
    "group": "Dev",
    "id": "achievement.dev.write-test",
    "experience": 43,
    "charisma": 42,
    "talent": 46,
    "strength": 44,
    "intellect": 45
  },
  {
    "group": "Dev",
    "id": "achievement.dev.write-test",
    "experience": 43,
    "charisma": 42,
    "talent": 46,
    "strength": 44,
    "intellect": 45
  }
]